keps/sig-apps/3939-include-terminating-pods-as-active/README.md

```diff
@@ -0,0 +1,781 @@
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Job API Definition
+
+At the JobSpec level, we are adding a new BoolPtr field:
+
+```golang
+type JobSpec struct{
+  ...
+ // terminatingAsActive specifies if the Job controller should include terminating pods
+ // as active. If the field is true, then the Job controller will include active pods
+ // to mean running or terminating pods
+ // +optional
+ TerminatingAsActive *bool
+}
+```
+
+### Deployment/ReplicaSet API
+
+```golang
+// DeploymentStrategy stores information about the strategy and rolling-update
+// behavior of a deployment.
+type DeploymentStrategy struct {
+  ... 
+  // TerminatingAsActive specifies if the Deployments should include terminating pods
+ // as active. If the field is true, then the Deployment controller will include active pods
+ // to mean running or terminating pods
+ // +optional
+ TerminatingAsActive *bool
+}
+```
+
+In [Option for acknowledging terminating Pods in Deployment rolling update](https://github.com/kubernetes/kubernetes/issues/107920)
+there was a request to add this as part of the `DeploymentStrategy` field.  Generally, handling terminating pods as active can be useful in both RollingUpdates and Recreating rollouts.  Having this field for both strategies allows for handling of terminating pods in both cases.  
+
+Deployments create ReplicaSets so there is a need to add a field in the ReplicaSet as well.  Since ReplicaSets are not typically
+set by users, we should add a field to the ReplicaSet that is set from the DeploymentSpec.  
+
+```golang
+// ReplicaSetSpec is the specification of a ReplicaSet.
+// As the internal representation of a ReplicaSet, it must have
+// a Template set.
+type ReplicaSetSpec struct {
+  ...
+ // TerminatingAsActive specifies if the Deployments should include terminating pods
+ // as active. If the field is true, then the Deployment controller will include active pods
+ // to mean running or terminating pods
+ // +optional
+ TerminatingAsActive *bool
+}
+```
+
+### Implementation
+
+Generally, both the Job controller and ReplicaSets utilize `FilterActivePods` in their reconciliation loop.  `FilterActivePods` gets a list of pods that are not terminating.  This KEP will include terminating pods in this list.
+
+```golang
+// FilterActivePods returns pods that have not terminated.
+func FilterActivePods(pods []*v1.Pod, terminatingPods bool) []*v1.Pod {
+ var result []*v1.Pod
+ for _, p := range pods {
+  if IsPodActive(p) {
+   result = append(result, p)
+  } else if IsPodTerminating(p) && terminatingPods {
```

Comment:
Spec IsPodTerminating here (so i can review) :)

Reply:
Updated this with `IsPodTerminating`